### ISSUE TITLE
Is it possible to highlight text on the next page?

### ISSUE THREAD

**Issue by Galunid** [koreader/koreader#4496]
I was wondering whether it is possible at all, to have koreader "turn page" when we highlight and hold a finger on the bottom right corner of the page, so that we could allow highlighting text, that is split between two pages. If it's not possible yet, are there any reasons that could prevent implementing this? If it is possible to implement, where can I find code responsible for drawing highlights?

**Comment by Frenzie** [comment]
The frontend code is here: https://github.com/koreader/koreader/blob/6ec3143f2f29df22a7eb472c4a0fb9da0039ef64/frontend/apps/reader/modules/readerhighlight.lua

I don't think I've ever really looked at it much.

**Comment by poire-z** [comment]
Not possible currently (already requested in #2071 and https://github.com/koreader/koreader/issues/4421#issuecomment-450661008), and the last times I looked at the code while thinking about that, it looked very complicated, and may involve work on each of the engines wrapping code (would need to turn page, render new page, deal with gesture state across these, build some combined text selection...)
The easiest workaround is to just switch to Scroll mode, which allows panning the page to bring start and end into view, then highlight, then switch back to Page mode. (But if you're text doesn't fit on a single page, this won't be enough).
(Or, intermediate solution: have koreader automatically switch to Scroll mode when text selection/panning reaches the bottom, or bottom right corner - while remembering and re-showing currently made selection - and allow scrolling while still panning - which still looks quite complicated, but could probably be made purely on the Lua side.)

**Comment by Galunid** [comment]
```diff
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -351,6 +351,16 @@ function ReaderHighlight:onHoldPan(_, ges)
             self.selected_word = nil
         end
     end
+    if self.holdpan_pos.x > 480 and self.holdpan_pos.y > 530 then
+        logger.dbg("Switched to scroll mode")
+        self.ui:handleEvent(Event:new("SetViewMode", "scroll"))
+        UIManager:scheduleIn(0.1, function()
+            logger.dbg("Moving to new position", self.ui.rolling.doc_height)
+            self.ui:handleEvent(Event:new("Panning",{ "north", 10 }, _))
+            self.ui:handleEvent(Event:new("SetViewMode", "page"))
+        end)
+    end
     UIManager:setDirty(self.dialog, "ui")
 end
```
So after playing around for a bit, that's what I've wrote. It... does scroll the page, when on the border, but the problem is, it doesn't update highlight starting position(at least that's what I think). It seems they are based off screen's coordinates, rather than book's location. Any ideas how to deal with it?

**Comment by poire-z** [comment]
That's a good start :) The switch to scroll mode is quite smooth.
But I think you want to stay in scroll mode, and only when the Hold is released and the selection made, or the higlight dialog closed, switch back to Page mode. Some flag of this temporary change of mode might be needeed.

You'll have to read some other modules to understand all that's involved, especially readerview.lua where you'll find some `ReaderView:screenToPageTransform(pos)` and `ReaderView:pageToScreenTransform(page, rect)`.

Might be easier to first try making that select & scroll & continue selecting directly in scroll mode.

(You're testing on EPUB I guess, which is a "readerrolling" thing, with a super huge height that you indeed can scroll and pages are just arbitrary slices of it - might be quite different and more complicated with PDF - so you might have to make, at the end, that stuff only available when `not self.ui.document.info.has_pages` (the classic test for EPUBs).)

**Comment by Galunid** [comment]
```01/19/19-19:56:04 DEBUG getTextFromPositions, position:  {
    [1] = {
        ["page"] = 12,
        ["x"] = 104,
        ["w"] = 0,
        ["y"] = 266,
        ["h"] = 0
    },
    [2] = {
        ["page"] = 14,
        ["x"] = 469,
        ["w"] = 0,
        ["y"] = 551,
        ["h"] = 0
    }
}
```
Is it possible that `getTextFromPositions` method ignores page attribute of position?
Looking at this log, 2 pages worth of text should be selected, while there is only a few sentences.

**Comment by Galunid** [comment]
So after some digging in
```lua
self.ui.document:getTextFromPositions(self.hold_pos, self.holdpan_pos)
```
Results in calling 
```lua
local text_range = self._document:getTextFromPositions(pos0.x, pos0.y, pos1.x, pos1.y) -- it seems, that page is omnited here
```
Which is probably executed in `cre.cpp`
```cpp
static int getTextFromPositions(lua_State *L) {
	CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
	int x0 = luaL_checkint(L, 2); // seems like we only check x0, x1, y0, y1 onscreen(?) coordinates
	int y0 = luaL_checkint(L, 3);
	int x1 = luaL_checkint(L, 4);
	int y1 = luaL_checkint(L, 5);

	LVDocView *tv = doc->text_view;
	lvRect margin = tv->getPageMargins();

	lvPoint startpt(x0, y0);
	lvPoint endpt(x1, y1);
	ldomXPointer startp = tv->getNodeByPoint(startpt);
	ldomXPointer endp = tv->getNodeByPoint(endpt);
	if (!startp.isNull() && !endp.isNull()) {
	    lua_newtable(L); // new text boxes
		ldomXRange r(startp, endp);
		if (r.getStart().isNull() || r.getEnd().isNull())
			return 0;
		r.sort();

		// When panning, only extend to include the whole word when we are
		// actually holding inside a word. This allows selecting punctuation,
		// quotes or parens at start or end of selection to have them
		// included in the highlight.
		bool not_panning = r.getStart() == r.getEnd();
		if ( (not_panning || r.getStart().isVisibleWordChar()) && !r.getStart().isVisibleWordStart())
			r.getStart().prevVisibleWordStart();
		if ( (not_panning || r.getEnd().isVisibleWordChar()) && !r.getEnd().isVisibleWordEnd())
			r.getEnd().nextVisibleWordEnd();
		if (r.isNull())
			return 0;

		if (r.getStart() == r.getEnd()) { // for single CJK character
			ldomNode * node = r.getStart().getNode();
			lString16 text = node->getText();
			int textLen = text.length();
			int offset = r.getEnd().getOffset();
			if (offset < textLen - 1)
				r.getEnd().setOffset(offset + 1);
		}

		r.setFlags(1);
		tv->selectRange(r);  // we let crengine do native highlight of selection

		/* We don't need these:
		int page = tv->getBookmarkPage(startp);
		int pages = tv->getPageCount();
		lString16 titleText;
		lString16 posText;
		tv->getBookmarkPosText(startp, titleText, posText);
		*/
		lString16 selText = r.getRangeText( '\n', 8192 );

		lua_pushstring(L, "text");
		lua_pushstring(L, UnicodeToLocal(selText).c_str());
		lua_settable(L, -3);
		lua_pushstring(L, "pos0");
		lua_pushstring(L, UnicodeToLocal(r.getStart().toString()).c_str());
		lua_settable(L, -3);
		lua_pushstring(L, "pos1");
		lua_pushstring(L, UnicodeToLocal(r.getEnd().toString()).c_str());
		lua_settable(L, -3);
		/* We don't need these:
		lua_pushstring(L, "title");
		lua_pushstring(L, UnicodeToLocal(titleText).c_str());
		lua_settable(L, -3);
		lua_pushstring(L, "context");
		lua_pushstring(L, UnicodeToLocal(posText).c_str());
		lua_settable(L, -3);
		lua_pushstring(L, "percent");
		lua_pushnumber(L, 1.0*page/(pages-1));
		lua_settable(L, -3);
		*/
	    return 1;
	}
    return 0;
}
```

**Comment by poire-z** [comment]
```c
int x0 = luaL_checkint(L, 2); // seems like we only check x0, x1, y0, y1 onscreen(?) coordinates
lvPoint startpt(x0, y0);
ldomXPointer startp = tv->getNodeByPoint(startpt)
```
The last call uses lvdocview.cpp `LVDocView::windowToDocPoint(lvPoint & pt)` and indeed delegates that work of screen position to document position conversion to crengine as it knows what's displayed in which mode. You surely don't want to mess with that :)

`getTextFromPositions(lua_State *L)` also let crengine do `native highlight of selection` so the Lua code doesn't have to.

My feeling is that you shouldn't need to mess with cre.cpp and crengine code as long as your selection is on a single screen page (if you switch to scroll mode): everything could probably be done with some coordinate works in Lua. Like updating hold_pos to reflect the page y-shift or cleaning up and simulating the initial hold position, stuff like that.

(If you were to want it to work on 3+-multi-pages, as you get XPointers (strings like xpath `/body/div[2]/span[4]`), you could keep the startXPointer even if it goes off screen, and get the whole text by adding a getTextFromXPointer**s**() alongside the existing `getTextFromXPointer` like there is a getHTMLFromXPointer**s** .)

**Comment by Galunid** [comment]
```diff
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -11,8 +11,20 @@ local logger = require("logger")
 local _ = require("gettext")
 local T = require("ffi/util").template
 
+local scrolled_steps = -3
 local ReaderHighlight = InputContainer:new{}
 
+
+local function copy(obj, seen)
+  if type(obj) ~= 'table' then return obj end
+  if seen and seen[obj] then return seen[obj] end
+  local s = seen or {}
+  local res = setmetatable({}, getmetatable(obj))
+  s[obj] = res
+  for k, v in pairs(obj) do res[copy(k, s)] = copy(v, s) end
+  return res
+end
+
 function ReaderHighlight:init()
     self.ui:registerPostInitCallback(function()
         self.ui.menu:registerToMainMenu(self)
@@ -335,7 +347,9 @@ function ReaderHighlight:onHoldPan(_, ges)
     self.holdpan_pos = self.view:screenToPageTransform(ges.pos)
     logger.dbg("holdpan position in page", self.holdpan_pos)
     local old_text = self.selected_text and self.selected_text.text
-    self.selected_text = self.ui.document:getTextFromPositions(self.hold_pos, self.holdpan_pos)
+    hold_pos_new = copy(self.hold_pos)
+    hold_pos_new.y = hold_pos_new.y + scrolled_steps
+    self.selected_text = self.ui.document:getTextFromPositions(--[[self.hold_pos]]hold_pos_new, self.holdpan_pos)
     if self.selected_text and old_text and old_text == self.selected_text.text then
         -- no modification
         return
@@ -350,6 +364,19 @@ function ReaderHighlight:onHoldPan(_, ges)
             #self.selected_text.sboxes > 1 then
             self.selected_word = nil
         end
+        local is_in_corner = self.holdpan_pos.x > 400 and self.holdpan_pos.y > 450
+        local scroll = self.view.view_mode ~= "scroll"
+        logger.dbg("scroll_mode?", scroll)
+        if not self.ui.document.info.has_pages and is_in_corner then
+            if self.view.view_mode ~= "scroll" then
+                self.ui:handleEvent(Event:new("SetViewMode", "scroll"))
+                scrolled_steps = -3
+            end
+--            UIManager:scheduleIn(0.1, function()
+                self.ui:handleEvent(Event:new("Panning",{ "north", 1 }, _))
+                scrolled_steps = scrolled_steps - self.ui.rolling.panning_steps.normal
+--            end)
+        end
     end
     UIManager:setDirty(self.dialog, "ui")
 end
@@ -626,6 +653,9 @@ function ReaderHighlight:onHoldRelease()
             self:lookupWikipedia()
             self:onClose()
         end
+
+        if self.view.view_mode == "scroll" then self.ui:handleEvent(Event:new("SetViewMode", "page")) end
+        scrolled_steps = -3
     elseif self.selected_word then
         self:lookup(self.selected_word, self.selected_link)
         self.selected_word = nil
```
This is what I've got so far. It works... more or less. The problem is that switching to "scroll" mode results in inconsistent page transformation, and I don't know how to determine the number of steps needed to compensate for this. Any ideas what could cause it?

**Comment by poire-z** [comment]
You seem to get something like `selected_text["pos0"] = "/body/DocFragment/body/div/p[15]/text()[9].50". a xpointer.
`self.ui.document:getPosFromXPointer(self.selected_text.pos0)` could give you : 3430 27 (y, then x), y in the full document height.

There's some code that guess the screen position from a xpointer at:
https://github.com/koreader/koreader/blob/e1cb983cec9a194a1937091b76a3643a93c8d80a/frontend/apps/reader/modules/readerlink.lua#L289-L301

Looks like the shift between scroll and page mode is just `doc_margins["top"] + header_height `
I guess that doing that from the same pos0 each time you scroll could give you the scroll distance..

> is there any function responsible for copying object?

We don't seem to have one. But your hold_pos is a generic table of 5 elements (page, x, y, w, h). You may as well just create a new table and set these 5 items from the old one.

**Comment by Frenzie** [comment]
You could also use a more generic utility function like this:

https://github.com/koreader/koreader/pull/3176/commits/e83f3bfbb92fbe3d1e6738a00ba55938b4979141

PS That was never mainlined in case you're wondering. I only removed it from the PR.

**Comment by arooni** [comment]
I'd also love to see this feature.  One of the biggest things I miss from the stock kindle experience (the other being adding notes to a highlighted selection)

**Comment by Galunid** [comment]
I just tested it on my kindle and it works pretty well. The only problem is that it sometimes doesn't detect when my finger is in the corner. I've changed `15/16*Screen:getWidth()` to `14/16*Screen:getWidth()`, as it didn't always detect that my finger was in the corner. With this change we should probably check whether user is in landscape or portrait orientation, and swap the coefficients. Good idea with scrolling 1/3 of the page height, solves a lot of strange issues that I encountered when scrolling line by line.
